gem/optimise.py

```diff
@@ -413,3 +474,758 @@ def aggressive_unroll(expression):
+    Fields:
+    1. node
+    2. multiindex
+    3. arg_ind
+    4. rep: a list of dictionaries (summands) of list of factors, such that
+    node = Sum(summands), where a summand is Product(factors), possibly contracted
+    with multiindex
+    Factors are indexed with keys:
+        1. One of the argument indices (e.g. j):
+            factors depend on j but not other argument indices
+            All summands will have this item (to avoid checking existence),
+            with the values possibly as empty list
+        2. Combination of >1 argument indices (e.g. j, k):
+            factors depend on j and k, but not other argument indices
+            These are rarer and are added as factors are encourtered
+        3. string 'const':
+            factors with no free indices
+        4. string 'quad':
+            factors depend on indices (typically quadrature indices for reduction)
+            other than argument indices
+    """
+    def __init__(self, node, arg_ind, rep=None):
+        """
+        Constructor
+        :param node: gem node
+        :param arg_ind: tuples of tuples of argument (linear) indices
+        """
+        self.node = node
+        self.arg_ind = arg_ind
+        # TODO: make these two properties of the object
+        self.arg_ind_flat = tuple([i for id in self.arg_ind for i in id])
+        self.arg_ind_set = set(self.arg_ind_flat)
+        if isinstance(node, IndexSum):
+            self.multiindex = node.multiindex
+            self.node = node.children[0]
+        else:
+            self.multiindex = ()
+        if rep:
+            self.rep = rep
+        else:
+            self._build_repr()
+
+
+    def _decide_key(self, factor):
+        """
+        Helper function to decide the appropriate key of a factor
+        :param factor: gem node
+        """
+        fi = factor.free_indices
+        if not fi:
+            return 'const'
+        else:
+            ind_set = set(fi) & arg_ind_set
+            if len(ind_set) > 1:
+                return tuple(i for i in arg_ind_flat if i in ind_set)
+            elif len(ind_set) == 0:
+                return 'quad'
+            else:
+                return ind_set.pop()
+
+    def _build_repr(self):
+        """
+        Build self.repr from self.node
+        """
+        node = expand_products(self.node, self.arg_ind_flat)
+        summands = collect_terms(node, Sum)
+        rep = []
+        for summand in summands:
+            d = OrderedDict()
+            for i in ['const', 'quad'] + list(self.arg_ind_flat):
+                d[i] = list()
+            for factor in collect_terms(summand, Product):
+                key = self._decide_key(factor)
+                if key in d:
+                    d[key].append(factor)
+                else:
+                    d[key] = [factor]
+            rep.append(d)
+        self.rep = rep
+
+    def find_optimal_arg_factors(self):
+        # TODO: add guard so that self.rep is up to date here
```

Comment:
Yes you are right, self.rep is always latest. I should worry about self.opt_node.